### PR TITLE
Support for channels

### PR DESCRIPTION
**Pull request by megaadam**
@TobbeEdgeware @TSonono @nilsandgren @lithammer 
Enjoy!

It seems after all that you are supposed merge upstream from your own fork.

### CODE REVIEW

**Comment by megaadam** [comment]
It would be nice to hear your views before I PR to grafov

File: M3U8.md [hunk 1]
```diff
@@ -83,14 +83,14 @@ Legend for playlist types:
 IETF drafts notes
 -----------------
 
-[IETF](http://ietf.org) document currently in Draft status. Different versions of the document introduce changes of HLS protocol playlist formats. Latest version of the HLS protocol is version 7.
```
Comment (TobbeEdgeware) [inline M3U8.md]: Why do this crappy updates on HLS version which has nothing to do with Channels?

If you do this, do this properly. The latest draft of HLS is https://datatracker.ietf.org/doc/html/draft-pantos-hls-rfc8216bis-10 and the latest version of HLS is 10.

I'd suggest to just drop this and the next changes to the repo.

File: reader.go [hunk 2]
```diff
@@ -329,6 +329,12 @@ func decodeLineOfMasterPlaylist(p *MasterPlaylist, state *decodingState, line st
 				alt.Subtitles = v
 			case "URI":
 				alt.URI = v
+			case "CHANNELS":
+				channels, err := strconv.Atoi(v)
+				if err != nil {
+					return fmt.Errorf("non-integer value \"%s\" for CHANNELS attribute", v)
```
Comment (TobbeEdgeware) [inline reader.go]: use %q instead of `"%s"`

File: writer.go [hunk 3]
```diff
@@ -156,6 +156,11 @@ func (p *MasterPlaylist) Encode() *bytes.Buffer {
 					p.buf.WriteRune('"')
 				}
 				p.buf.WriteRune('\n')
+				if alt.Channels != 0 {
```
Comment (TobbeEdgeware) [inline writer.go:158]: This must go before the newline just above.

File: sample-playlists/from-apple/master.m3u8 [hunk 4]
```diff
@@ -0,0 +1,152 @@
+#EXTM3U
```
Comment (TobbeEdgeware) [inline sample-playlists/from-apple/master.m3u8]: why this extremely long list to check for one simple attribute?
And why remove media-playlist-large and add media-playlist-long instead?

File: reader_test.go [hunk 5]
```diff
@@ -972,6 +972,64 @@ func TestDecodeMediaPlaylistStartTime(t *testing.T) {
...
+	if alt2.Type != "AUDIO" {
+		t.Error("Expected AUDIO track in test input Alternatives[2]")
+	}
+
+	if alt2.Channels != 6 {
+		t.Error("Expected 6 channels track in test input Alternatives[2]")
+	}
+}
+
+func TestDecodeMediaPlaylistLong(t *testing.T) {
```
Comment (TobbeEdgeware) [inline reader_test.go]: What does this test have to do with channels?

**Review by TobbeEdgeware — COMMENTED** [review]
This is way too many changes and to just add CHANNELS to the master playlist.

Especially if this should be submitted as a PR it must be focused on the one thing that it addresses and make it without big changes to any files.

File: reader_test.go (hunk 5 above)
Comment (megaadam) [reply reader_test.go]: Nothing, technically, but:
the other Channels-PR https://github.com/grafov/m3u8/pull/168 had an increase of the Playlist size. I was making sure that was not necessary

**Review by TobbeEdgeware — APPROVED** [review]
LGTM 👍

**Comment by megaadam** [comment]
I should also add `CHANNEL` tests for the `writer`. But I need to merge and move on now.

**Pull request merged**